docs/design/kmp-nversions.md

```diff
@@ -0,0 +1,106 @@
+      - name: yourCertName
+        maxVersionCount: 3
+    tenantID:
+    clientID:
+```
+
+Example: AKV KMP resource status with multiple versions retained in the cache
+
+```yaml
+Status:
+  Issuccess:        true
+  Lastfetchedtime:  2024-10-02T14:58:54Z
+  Properties:
+    Certificates:
+      Last Refreshed:  2024-10-02T14:58:54Z
+      Name:            yourCertName
+      Version:         a1b2c3d4e5f67890abcdef1234567890
+      Status:          Enabled
+      Last Refreshed:  2024-10-02T14:58:54Z
+      Name:            yourCertName
+      Version:         0ff373a9259c4578a247cfd7861a8805
+      Status:          Disabled
+```
+
+## Implementation Details
+
+- Modify the KMP data structure to include the status of the version.
+    ```go
+    type KMPMapKey struct {
+    Name    string
+    Version string
+    Status  string // Enabled or Disabled
+    Created time.Time // Time the version was created used for determining the oldest version
+    }
+    ```
+- Add the `maxVersionCount` parameter to the KMP resource in Ratify.
+  - ensure the value cannot be less than 1 or a negative number
+  - default to 0 if not specified
+  - maximum value should be (TBD)
+  - specify the value at the object level within the parameters of the KMP resource.
+- Changes to `azurekeyvault` provider:
+   - support for the `maxVersionCount` parameter.
+   - allowing retrieval of multiple versions of certificates or keys.
+   - remove the oldest version from the cache when the number of versions exceeds the `maxVersionCount` parameter.
+   - update disabled certs status in the cache & remove the certData from the cache.
+- Log when the status of a version changes.
+- Log when a conflict between the `maxVersionCount` and the number of specified certificate versions occurs.
+
+## Dev Work Items
+
+## Open Questions
+
+- If a version is disabled, should it be removed from the cache or retained based on the nVersionCount and marked as inactive\disabled?
+  - [x] Keep the disabled version in the cache and mark it as disabled.
+- If a version is disabled, does that count towards the nVersionCount? For example, if nVersionCount is set to 3 and one of the versions is disabled, should Ratify retain the last three active versions or the last three versions, regardless of their status?
+  - [x] Yes, disabled versions should count towards the nVersionCount. The reason for this is that disabled versions may be re-enabled in the future, and it is important to retain them in the cache.
+- Should the existing KMP data structure be changed to group versions by key or certificate name?
+  - [x] No, a flat list of versions is sufficient. At this time, there is no need to group versions by key or certificate name because the verifiers do not need to know the history of the versions.
+- Should the KMP status return a flat list of versions?
+  - [x] Yes, the status should return a flat list of versions.
+- What should the maximum value for nVersionCount be?
+  - [ ] TBD
```

Comment:
Maybe we can start with `2`, which means we will cache the latest version and one previous version.

Reply:
That does break away from the pattern secret store driver has, but I'm not opposed. 

It does bring up an interesting point, what exactly is the int counting for the value of the history? Is it the total number of the versions stored in the KMP or latest +n? for example the secret store, always stores latest and then if you specify to store one more than latest you would pass `1` to the `objectVersionHistory` option.

Reply:
discussed during community meeting, versionHistory defaults to 2  （ keeping 3 version)

Reply:
LGTM

Reply:
> versionHistory

LGTM